File: simulators/clrty_mirror/telemetry.py
"""Unified telemetry schema for ABM simulation."""
from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any, Iterator


@dataclass
class StateSnapshot:
    mid_price: float
    bid_depth_usdc: float
    ask_depth_usdc: float
    staked_nano: float = 0.0
    entropy_root: str = ""


@dataclass
class TelemetryLog:
    timestamp: str
    event_type: str
    event_id: int
    pre_state: StateSnapshot
    post_state: StateSnapshot
    status: str
    event_hash: str
    agent_actions: list[dict[str, Any]]
    chain_id: str = "clrty-1"


def read_ndjson(path: Path) -> Iterator[dict[str, Any]]:
    if not path.is_file():
        return iter(())
    for line in path.read_text().splitlines():
        line = line.strip()
        if line:
            yield json.loads(line)


def read_ticks(path: Path) -> list[dict[str, Any]]:
    if not path.is_file():
        return []
    return json.loads(path.read_text())


def read_telemetry(path: Path) -> list[dict[str, Any]]:
    return list(read_ndjson(path))


def write_telemetry_log(path: Path, log: TelemetryLog) -> None:
    with path.open("a") as f:
        f.write(json.dumps(asdict(log)) + "\n")

File: simulators/state_space/src/telemetry.rs
//! Unified simulation telemetry — pre/post state JSON for every event.

use crate::abm::AgentAction;
use crate::event_loop::{SimEventResult, SimExecutionStatus};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimStateSnapshot {
    pub mid_price: f64,
    pub bid_depth_usdc: f64,
    pub ask_depth_usdc: f64,
    pub staked_nano: u64,
    pub entropy_root: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimTelemetryLog {
    pub timestamp: String,
    pub event_type: String,
    pub event_id: u16,
    pub pre_state: SimStateSnapshot,
    pub post_state: SimStateSnapshot,
    pub status: String,
    pub event_hash: String,
    pub agent_actions: Vec<AgentAction>,
    pub chain_id: String,
}

impl SimTelemetryLog {
    pub fn from_event(
        pre: SimStateSnapshot,
        post: SimStateSnapshot,
        result: &SimEventResult,
        agents: Vec<AgentAction>,
        event_type: &str,
    ) -> Self {
        let status = match result.execution.status {
            SimExecutionStatus::Filled => "filled",
            SimExecutionStatus::Rejected => "rejected",
            SimExecutionStatus::Aborted => "aborted",
        };
        Self {
            timestamp: result.timestamp.clone(),
            event_type: event_type.to_string(),
            event_id: result.event_id,
            pre_state: pre,
            post_state: post,
            status: status.to_string(),
            event_hash: result.event_hash.clone(),
            agent_actions: agents,
            chain_id: result.chain_id.clone(),
        }
    }
}
